### PR TITLE
Adding du for composing chainables

### PR DESCRIPTION
**Pull request by pfgray**
`du` is a translation of scala's `for` comprehension for javascript.

It allows you to convert nested `chain` + `map` calls to top-level functions with increasing arity:
 
```ts
import { option, some, Option } from "fp-ts/lib/Option";

//from:
some("bob").chain(str => {
  return some(54).map(num => {
    return ({ name: str, age: num })
  })
})

//into:
du(option)(
  some("bob"),
  (str) => some(54),
  (num, str) => ({ name: str, age: num })
);
```

### CODE REVIEW

**Comment by gcanti** [comment]
Hi @pfgray

A positional API looks a bit awkward in case of refactorings, what about a builder instead?

```purescript
do
  name <- some('bob')
  age <- some(54)
  len <- pure name.length
  return [name, age, len]
```  

```ts
// user: Option<[string, number, number]>

const user = Do(option)
  .bind('name', () => some('bob'))
  .bind('age', () => some(54))
  .bind('len', ({ name }) => some(name.length))
  .return({ name, age, len } => [name, age, len])
```

where (pseudocode)

```ts
bind(this: Builder<S>, name: string, value: (scope: S) => Option<A>): Builder<S & { [name]: A }>
``` 

/cc @mattiamanzati @joshburgess

**Comment by gcanti** [comment]
I've done something similar in `fp-ts-routing`

- https://github.com/gcanti/fp-ts-routing/blob/fe5789cffeda104434c2918cf99b08aad168eec7/src/index.ts#L116
- https://github.com/gcanti/fp-ts-routing/blob/fe5789cffeda104434c2918cf99b08aad168eec7/src/index.ts#L96

**Comment by pfgray** [comment]
Ooh I like this... Will play around with this tonight

**Comment by mattiamanzati** [comment]
@gcanti I'd also Omit the key you're binding from the types, so you are sure that subsequent bind calls will override previous scope items.

**Comment by gcanti** [comment]
@mattiamanzati I'd argue that it should be the other way around if possible: if you try to bind the same name twice it should raise a static error

**Comment by sledorze** [comment]
@gcanti agree and I think you already wrote a type level poc of that behavior actually..

**Comment by gcanti** [comment]
@sledorze more than a POC actually, the second link I posted points to a possible implementation (the `RowLacks` type)

**Comment by sledorze** [comment]
Oh yes :)

**Comment by pfgray** [comment]
@gcanti I'm loving the builder:

bind

One small note,

I think we can only do one of three:

1. Generalize for `Chain`:
```
Do(option, 'name', some("foo"))
  .return(({name}) => ...)
```

2. Generalize for `Monad`:
```
Do(option)
  .return(({}) => ...) // here we can just `F.of({})`
```

3. Generalize for `Chain`:
```
Do(option)
  // can't call `return` because nothing is "bound"

Do(option)
  .bind('name', some("foo")) // must bind first, then subsequent binds take a function & can 'return'
  .bind('len', ({name}) => some(name.len))
  .return(({name, len}) => ...)
```

1 & 2 are pretty simple, and 3 would require an additional `DoBuilder` type (probably worth it, IMO)

Alternatively, we could name the first bind something like 'init' to distinguish it with something like:
```
Do(option)
  .init('name', some("foo")) // 'init' must be called first
  .bind('len', ({len}) => ...)
```

**Comment by gcanti** [comment]
@pfgray Here's some typings so we can reason about the methods / signatures (names are temporary)

```ts
import { Monad1 } from 'fp-ts/lib/Monad'
import { URIS, Type } from 'fp-ts/lib/HKT'

export type RowLacks<O extends object, K extends string> = O & Record<Extract<keyof O, K>, never>

interface Do1<M extends URIS, S extends object> {
  bind: <N extends string, A>(
    this: Do1<M, RowLacks<S, N>>,
    name: N,
    ma: Type<M, A>
  ) => Do1<M, S & { [K in N]: A }>
  bindL: <N extends string, A>(
    this: Do1<M, RowLacks<S, N>>,
    name: N,
    f: (s: S) => Type<M, A>
  ) => Do1<M, S & { [K in N]: A }>
  return: <A>(f: (s: S) => A) => Type<M, A>
}

declare function Do<M extends URIS>(M: Monad1<M>): Do1<M, {}>

//
// Usage
//

import { option, some } from 'fp-ts/lib/Option'

const user = Do(option)
  .bind('name', some('bob'))
  .bindL('age', () => some(54))
  // .bindL('age', () => some(54)) // dupicated binding, static error if decommented
  .bindL('len', ({ name }) => some(name.length))
  .return(({ name, age, len }) => [name, age, len] as const)
```

**Comment by mattiamanzati** [comment]
It would be also great to have a method that chaines a computation without storing the resulting value in the object

**Comment by pfgray** [comment]
@gcanti Good idea,

I guess I was trying to figure out if it was worth it to make `Do` generalized on `Chain` instead of `Monad`. in order to allow this, a caller must bind some value first, in order to have a chainable to chain off of:

```
import { option, some } from 'fp-ts/lib/Option'

type DoInitBuilder1<C extends URIS> = {
  bind<N extends string, A>(name: N, a: Type<C, A>): DoBuilder1<C, Record<N, A>>
}
type DoBuilder1<C extends URIS, A extends object> = {
  bind<N extends string, B>(this: DoBuilder1<C, RowLacks<A, N>>, name: N, c: Type<C, B>): DoBuilder1<C, A & Record<N, B>>
  bindL<N extends string, B>(this: DoBuilder1<C, RowLacks<A, N>>, name: N, f: (a: A) => Type<C, B>): DoBuilder1<C, A & Record<N, B>>
  return<Z>(f: (a: A) => Z): Type<C, Z>
}

declare function Do<C extends URIS>(C: Chain1<C>): DoInitBuilder1<C>

Do(option)
  .bind('name', some('bob')) // bind must be called first!
  .bindL('age', () => some(54))
  // .bindL('age', () => some(54)) // duplicated binding, static error if decommented
  .bindL('len', ({ name }) => some(name.length))
  .return(({ name, age, len }) => [name, age, len])
```

**Comment by gcanti** [comment]
@mattiamanzati so a `do/doL` pair?

```ts
interface Do1<M extends URIS, S extends object> {
  do: (ma: Type<M, unknown>) => Do1<M, S>
  doL: (f: (s: S) => Type<M, unknown>) => Do1<M, S>
  bind: <N extends string, A>(
    this: Do1<M, RowLacks<S, N>>,
    name: N,
    ma: Type<M, A>
  ) => Do1<M, S & { [K in N]: A }>
  bindL: <N extends string, A>(
    this: Do1<M, RowLacks<S, N>>,
    name: N,
    f: (s: S) => Type<M, A>
  ) => Do1<M, S & { [K in N]: A }>
  return: <A>(f: (s: S) => A) => Type<M, A>
}

const user = Do(option)
  .bind('name', some('bob'))
  .bindL('age', () => some(54))
  .doL(() => some(true)) // doesn't affect the scope
  .do(some(false)) // doesn't affect the scope
  // .bindL('age', () => some(54)) // duplicated binding, static error if decommented
  .bindL('len', ({ name }) => some(name.length))
  .return(({ name, age, len }) => [name, age, len] as const)
```

@pfgray `Chain` alone (i.e. without an `of` operation) seems pretty rare. Actually I can't think of an example... do you know a counterexample?

**Comment by pfgray** [comment]
Good point, can't think of any ATM

**Comment by gcanti** [comment]
@pfgray we can drop `RowLack` + `this` annotation in favour of a lightweight solution (using `Exclude`)

```ts
export interface Do1<M extends URIS, S extends object> {
  do: (ma: Type<M, unknown>) => Do1<M, S>
  doL: (f: (s: S) => Type<M, unknown>) => Do1<M, S>
  bind: <N extends string, A>(name: Exclude<N, keyof S>, ma: Type<M, A>) => Do1<M, S & { [K in N]: A }>
  bindL: <N extends string, A>(name: Exclude<N, keyof S>, f: (s: S) => Type<M, A>) => Do1<M, S & { [K in N]: A }>
  return: <A>(f: (s: S) => A) => Type<M, A>
  done: () => Type<M, S>
}
```

Moreover the static error is more localized (points to the offending LOC instead of the whole expression)

**Comment by anttih** [comment]
How is the inference for something like `Either`?

**Comment by gcanti** [comment]
@anttih good as long as we fix `L` after the first operation in `Do2` (for example by returning `Do2C`)

```ts
import { either, right } from 'fp-ts/lib/Either'
import { tuple } from 'fp-ts/lib/function'
import { HKT, Type, Type2, URIS, URIS2 } from 'fp-ts/lib/HKT'
import { Monad, Monad1, Monad2, Monad2C } from 'fp-ts/lib/Monad'

export interface Do2<M extends URIS2, S extends object> {
  do: <L>(ma: Type2<M, L, unknown>) => Do2C<M, S, L>
  doL: <L>(f: (s: S) => Type2<M, L, unknown>) => Do2C<M, S, L>
  bind: <N extends string, A, L>(name: Exclude<N, keyof S>, ma: Type2<M, L, A>) => Do2C<M, S & { [K in N]: A }, L>
  bindL: <N extends string, A, L>(
    name: Exclude<N, keyof S>,
    f: (s: S) => Type2<M, L, A>
  ) => Do2C<M, S & { [K in N]: A }, L>
  return: <A, L>(f: (s: S) => A) => Type2<M, L, A>
  done: <L>() => Type2<M, L, S>
}

export interface Do2C<M extends URIS2, S extends object, L> {
  do: (ma: Type2<M, L, unknown>) => Do2C<M, S, L>
  doL: (f: (s: S) => Type2<M, L, unknown>) => Do2C<M, S, L>
  bind: <N extends string, A>(name: Exclude<N, keyof S>, ma: Type2<M, L, A>) => Do2C<M, S & { [K in N]: A }, L>
  bindL: <N extends string, A>(
    name: Exclude<N, keyof S>,
    f: (s: S) => Type2<M, L, A>
  ) => Do2C<M, S & { [K in N]: A }, L>
  return: <A>(f: (s: S) => A) => Type2<M, L, A>
  done: () => Type2<M, L, S>
}

export interface Do1<M extends URIS, S extends object> {
  do: (ma: Type<M, unknown>) => Do1<M, S>
  doL: (f: (s: S) => Type<M, unknown>) => Do1<M, S>
  bind: <N extends string, A>(name: Exclude<N, keyof S>, ma: Type<M, A>) => Do1<M, S & { [K in N]: A }>
  bindL: <N extends string, A>(name: Exclude<N, keyof S>, f: (s: S) => Type<M, A>) => Do1<M, S & { [K in N]: A }>
  return: <A>(f: (s: S) => A) => Type<M, A>
  done: () => Type<M, S>
}

export interface Do0<M, S extends object> {
  do: (ma: HKT<M, unknown>) => Do0<M, S>
  doL: (f: (s: S) => HKT<M, unknown>) => Do0<M, S>
  bind: <N extends string, A>(name: Exclude<N, keyof S>, ma: HKT<M, A>) => Do0<M, S & { [K in N]: A }>
  bindL: <N extends string, A>(name: Exclude<N, keyof S>, f: (s: S) => HKT<M, A>) => Do0<M, S & { [K in N]: A }>
  return: <A>(f: (s: S) => A) => HKT<M, A>
  done: () => HKT<M, S>
}

export function Do<M extends URIS2>(M: Monad2<M>): Do2<M, {}>
export function Do<M extends URIS2, L>(M: Monad2C<M, L>): Do2C<M, {}, L>
export function Do<M extends URIS>(M: Monad1<M>): Do1<M, {}>
export function Do<M>(M: Monad<M>): Do0<M, {}>
export function Do<M>(_: Monad<M>): Do0<M, {}> {
  return null as any
}

// const either1: Either<string, [string, number, number]>
export const either1 = Do(either)
  .bind('name', right<string, string>('bob'))
  .bindL('age', () => right(54))
  // .bindL('bad', () => right<boolean, number>(54)) // static error Type 'boolean' is not assignable to type 'string' if decommented
  .doL(() => right(true)) // doesn't affect the scope
  .do(right(false)) // doesn't affect the scope
  // .bindL('age', () => right(54)) // duplicated binding, static error if decommented
  .bindL('len', ({ name }) => right(name.length))
  .return(({ name, age, len }) => tuple(name, age, len))
```

File: src/Do.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+export interface Do0<M, S extends object> {
+  do: (ma: HKT<M, unknown>) => Do0<M, S>
+  doL: (f: (s: S) => HKT<M, unknown>) => Do0<M, S>
+  bind: <N extends string, A>(name: Exclude<N, keyof S>, ma: HKT<M, A>) => Do0<M, S & { [K in N]: A }>
+  bindL: <N extends string, A>(name: Exclude<N, keyof S>, f: (s: S) => HKT<M, A>) => Do0<M, S & { [K in N]: A }>
+  return: <A>(f: (s: S) => A) => HKT<M, A>
+  done: () => HKT<M, S>
+}
+
+export function Do<M extends URIS2>(M: Monad2<M>): Do2<M, {}>
```
Comment (gcanti) [inline src/Do.ts]: Some missing overloads

```ts
export function Do<M extends URIS3>(M: Monad3<M>): Do3<M, {}>
export function Do<M extends URIS3, U, L>(M: Monad3C<M, U, L>): Do3C<M, {}, U, L>
```

**Comment by gcanti** [comment]
@pfgray I'd try to return `Do2C` from `Do2` (and `Do3C` from `Do3`). Current typings allow to mix left types, we can also add some type-level test in `dtslint/ts3.0/index.ts`

Example

```ts
//
// Do
//

import { either, right } from 'fp-ts/lib/Either'
import { Do } from '../../src/Do'

export const either1 = Do(either)
  .bind('name', right<string, string>('bob'))
  // $ExpectError
  .bindL('bad', () => right<boolean, number>(54))
```

**Comment by gcanti** [comment]
@pfgray you can fix travis by running `npm run fix-prettier`

**Comment by gcanti** [comment]
Thanks @pfgray !

**Comment by pfgray** [comment]
@gcanti , thanks for all your help!

**Comment by rektide** [comment]
@pfgray first of all: wahoo! 🤩

Would you mind walking me through some of the things going on here? I'd love to get a better general understanding of this code, but also I'm pretty new to a lot of the things going on here.

* Do the generic parameters stand for anything? `M`, `S`, `U`, `L`, `N` (name?), `K`, `A` are what I'm seeing, I think. How many of those have some kind of meaning? 
* What leads to there being the multiple different interfaces here (`Do0`, `Do1`, `Do2C`, `Do2`, `Do3C`, `Do3`)? Is there some relation between them?

I'm still just trying to get started understanding this structure. Anything you might offer could help. Thanks.

**Comment by pfgray** [comment]
@rektide

TBH, this is one of the more advanced usages of fp-ts, so bear with me, wall of text inc :)

At a high level, `Do` is helpful syntax for composing monads.

You may have experienced this pain already, when chaining together multiple promises:

```
fetchUser().then(user => {
  return fetchEnrollment(user.id).then(enrollment => {
    return fetchCourse(enrollment.id).then(course => {
      return { user, enrollment, course }
    })
  })
})
```

What's interesting about the "pyramid of doom" problem is that it isn't unique to just promises, but _all monads_.

Consider the `Array` monad:
```
users.flatMap(user => {
  return user.friends.flatMap(friend => {
    return friend.favorites;
  })
})
```

We could build specialized `Do` functions for both arrays and promises, but then we'd have to do this for every other monad (even ones we have yet to invent)!

This `Do` is generalized for _all_ monads, meaning that it doesn't just work for arrays and promises.

Here is some pseudocode for the above examples converted to`Do`:
```
Do(promise)
  .bind('user', fetchUser())
  .bindL('enrollment', ({ user }) => fetchEnrollment(user.id))
  .bindL('course', ({ enrollment }) => fetchCourse(enrollment.id))
  .done()

Do(array)
  .bind('user', users)
  .bindL('friend', ({user}) => user.friends)
  .bindL('favorite', ({friend}) => friend.favorites)
  .return(({favorite}) => favorite)
```

`.bind` will take care of calling the monadic `chain` on the value you supply, extracting the value inside, and "binding" it into an internal object so that you can destructure it in a subsequent `bindL` call. 



The rest is just implementation details :)

> How does `Do` know when to call `flatMap`, and when to call `then`? 

Fp-ts uses _typeclasses_ to achieve polymorphism (as opposed to having an interface a client implements), which means that the 'chain' and 'of' methods are defined on a separate object and passed in the beginning:

```
const arrayTypeclass: Monad1<A> = {
  of: function(a) { ... },
  chain: function(a, f) { ... }
}

Do(arrayTypeclass)
  ...
```
Now you can pass arrays to `bind`, and `bindL` calls in order to flatMap over them.



Yeah, some of the generic parameters stand for things: `M` is the "Monad" type, i.e. option, array, etc.
`N` is the name of the key you're binding a value to.

The others follow a few conventions I've seen elsewhere and in this repo. `A` is used for the type "inside" a Monad, i.e. as in `number` in ` Option<number>`. `L` is used as the "left" side of a monad with two type params (as in `string` in `Either<string, number>`. TBH I'm not sure about the name behind `U` :), but it's usually used as the third type parameter to a Monad with three type parameters.

The various `Do1`, `DoN` types all correlate to monads with `N` type params. So, `Do1`; `Option<?>`, `Do2`; `Either<?, ?>`, etc. Typescript doesn't exactly support higher kinded types, and the trick that fp-ts uses to express causes this split.

**Pull request merged**